### PR TITLE
doc/haskell: begin bringing back haskell documentation to the nixpkgs manual

### PR DESCRIPTION
**Pull request by sternenseemann**
###### Motivation for this change

Reopened  #126382 with a branch in NixOS/nixpkgs, so we can have PRs against this branch more easily.

@cdepillabout can you post our projected TOC in the comments here?

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: doc/languages-frameworks/haskell.section.md
Comment (teto) [inline doc/languages-frameworks/haskell.section.md:15]: latest being hackage (!= hackage) ?

Comment (teto) [reply doc/languages-frameworks/haskell.section.md:15]: it's explained later on so I understand that if the package is on stackage, then the version is not necessarily the latest ?

**Comment by teto** [comment]
nice to have document this. I feel like this can be improved over time without necessarily having all in one PR.
Do you intend to mention haskell.nix ? since it helps wroking around nixpkgs limitations ?

File: doc/languages-frameworks/haskell.section.md
Comment (sternenseemann) [reply doc/languages-frameworks/haskell.section.md:15]: Yeah, I want to explain this in more detail in the section “Package Set Generation”. Basically we use the latest version of every package from hackage by default unless it is part of stackage or there has been a pin manually set up. In those cases we use the manual version constraint or the one from stackage to get the version of the standard attribute and additionally generate an attribute `package-name_A_B_C_D` which points to the latest version from hackage.

**Comment by sternenseemann** [comment]
> nice to have document this. I feel like this can be improved over time without necessarily having all in one PR.
> Do you intend to mention haskell.nix ? since it helps wroking around nixpkgs limitations ?

My current stance on this is that I'd like to accurately describe our infrastructure as well as its limitation, but I wouldn't like to link / endorse a third party project that is mostly run by a single company. Also this will probably just end with people coming to nixpkgs for their haskell.nix issues.

**Comment by cdepillabout** [comment]
> cdepillabout can you post our projected TOC in the comments here?

We had a video chat about what sections we want in our documentation.

So far, these sections have been completed in this PR:

- overview and introduction
- haskellPackages.mkDerivation (documented all arguments)

Here are the sections that are left to do, along with who volunteered to write them:

- running cabal2nix and packaging your own packages (this is section in old manual as well?), callHackage (@maralorn)
- section on overriding (@sternenseemann)
- patch nix-doc to generate documentation from haskell.lib
- document dev environment stuff shell.nix generation, env passthru, shellFor, developPackage
- HLS (@maralorn)
- contributing (@cdepillabout during zurihac?)
  - fixing a broken package
  - regenerate hackage2nix, as well as how the package set is generated in general
  - what do maintainers do
- FAQs / How-To
- reference
  - how does NIX_GHC work? (@maralorn)
  - IFD helpers (callCabal2nix, callHackage, callHackageDirect, etc)

Keep in mind that this is a very rough list, so there is likely to be big changes and reorganization.

---------------------------------------------

Another important thing is that @maralorn suggested we follow the Divio documentation system: https://documentation.divio.com/

We should keep these four types of documentation in mind as we are writing this manual.

**Comment by sternenseemann** [comment]
We should [document the url capabilities of cabal2nix](https://github.com/NixOS/cabal2nix/issues/440)

**Comment by sternenseemann** [comment]
- [ ] FAQ: Why don't build-tool-depends work with cabal v2 command?

**Comment by sternenseemann** [comment]
https://github.com/NixOS/nixpkgs/pull/172094#issuecomment-1121940016

> Is this worth adding a comment to remind and explain?

File: doc/languages-frameworks/haskell.section.md
Comment (maralorn) [inline doc/languages-frameworks/haskell.section.md:13]:
```suggestion
first-class support for the default compiler (currently GHC 9.0.2) and usually
```

or maybe we remove this line? I feel like it will bite us over and over?

**Review by maralorn — APPROVED** [review]
Absolutely great. Can‘t wait to see this merged so that we can all elaborate on it.

**Comment by sternenseemann** [comment]
- [ ] The issue described in https://github.com/NixOS/nixpkgs/pull/182023 would make for a good FAQ entry.

File: doc/languages-frameworks/haskell.section.md
Comment (sternenseemann) [reply doc/languages-frameworks/haskell.section.md:13]: Thanks, updated references to 8.10.7. I think it's fine once we've merged since I at least always do a grep over the repo to find references to the old GHC version; e.g. module docs already reference ghc vresions today, so there's always something to update.

File: doc/languages-frameworks/haskell.section.md
Comment (sternenseemann) [reply doc/languages-frameworks/haskell.section.md:60]: Right!

File: doc/languages-frameworks/haskell.section.md
Comment (sternenseemann) [reply doc/languages-frameworks/haskell.section.md:439]: Thanks,  fixed!

File: doc/languages-frameworks/haskell.section.md
Comment (sternenseemann) [reply doc/languages-frameworks/haskell.section.md:707]: I've changed it to `markUnbroken` now, I think just documenting one of the two is better, since it implies deprecation of the other subtly.

File: doc/languages-frameworks/haskell.section.md
Comment (sternenseemann) [reply doc/languages-frameworks/haskell.section.md:450]: Noted.

File: doc/languages-frameworks/haskell.section.md
Comment (ConnorBaker) [inline doc/languages-frameworks/haskell.section.md:825]: The `C` in `Config` should be lowercase: https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/haskell-modules/lib/compose.nix#L167

Ran into this typo when I copy-pasted from the document today.

**Comment by cdepillabout** [comment]
Request to document `doJailbreak` (and explain why we use the term "jailbreaking"): https://github.com/NixOS/nixpkgs/issues/35278

File: doc/languages-frameworks/haskell.section.md
Comment (sternenseemann) [reply doc/languages-frameworks/haskell.section.md:448]: Done in 18ea31126f04dd9820430781ade74dd54f9824b5

File: doc/languages-frameworks/haskell.section.md
Comment (sternenseemann) [reply doc/languages-frameworks/haskell.section.md:757]: 343c9132a138a9931ff2a8e0d02e38acccd542eb

File: doc/languages-frameworks/haskell.section.md
Comment (sternenseemann) [reply doc/languages-frameworks/haskell.section.md:433]: I have added a TODO for this, since I can't muster the energy to do this right now. I think it'll go away anyways when a Package Set Generation section is added.

**Comment by ScottFreeCode** [comment]
> * [ ]  FAQ: Why don't build-tool-depends work with cabal v2 command?

In case this PR's comments get used for future reference, is this issue https://gist.github.com/ScottFreeCode/ef9f254e2dd91544bba4a068852fc81f ?

**Pull request merged**
